File: mobile/modules/safeparts-core/ios/SafepartsCoreModule.swift
import ExpoModulesCore

// UniFFI-generated Swift bindings live in ./uniffi/

public final class SafepartsCoreModule: Module {
  public func definition() -> ModuleDefinition {
    Name("SafepartsCore")

    Function("supportedEncodings") { () -> [String] in
      supportedEncodings()
    }

    Function("getVersion") { () -> String in
      getVersion()
    }

    AsyncFunction("splitSecret") { (secretB64: String, k: Int, n: Int, encoding: String, passphrase: String?) -> [String] in
      let k8 = UInt8(clamping: k)
      let n8 = UInt8(clamping: n)
      return try splitSecret(secretB64: secretB64, k: k8, n: n8, encoding: encoding, passphrase: passphrase)
    }

    AsyncFunction("combineShares") { (shares: [String], encoding: String, passphrase: String?) -> String in
      return try combineShares(shares: shares, encoding: encoding, passphrase: passphrase)
    }
  }
}

File: mobile/modules/safeparts-core/android/src/main/java/safeparts/core/SafepartsCoreModule.kt
package safeparts.core

import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import uniffi.safeparts_mobile_bridge.SafepartsException
import uniffi.safeparts_mobile_bridge.combineShares
import uniffi.safeparts_mobile_bridge.getVersion
import uniffi.safeparts_mobile_bridge.splitSecret
import uniffi.safeparts_mobile_bridge.supportedEncodings

class SafepartsCoreModule : Module() {
  override fun definition() = ModuleDefinition {
    Name("SafepartsCore")

    Function("supportedEncodings") {
      return@Function supportedEncodings()
    }

    Function("getVersion") {
      return@Function getVersion()
    }

    AsyncFunction("splitSecret") { secretB64: String, k: Int, n: Int, encoding: String, passphrase: String? ->
      try {
        splitSecret(secretB64, k.toUByte(), n.toUByte(), encoding, passphrase)
      } catch (e: SafepartsException) {
        throw Exception(e.message)
      }
    }

    AsyncFunction("combineShares") { shares: List<String>, encoding: String, passphrase: String? ->
      try {
        combineShares(shares, encoding, passphrase)
      } catch (e: SafepartsException) {
        throw Exception(e.message)
      }
    }
  }
}
